crates/staged-sync/tests/sync.rs

```diff
@@ -0,0 +1,319 @@
+        let mut chainspec: ChainSpec = clique.0.genesis().clone().expect("clique should be configured with a genesis").into();
+        let remote_genesis = SealedHeader::from(provider.remote_genesis_block().await.unwrap());
+
+        let mut local_genesis_header = Header::from(chainspec.genesis().clone());
+
+        let hardforks = chainspec.hardforks();
+
+        // set initial base fee depending on eip-1559
+        if let Some(0) = hardforks.get(&Hardfork::London) {
+            local_genesis_header.base_fee_per_gas = Some(EIP1559_INITIAL_BASE_FEE);
+        }
+
+        let local_genesis = local_genesis_header.seal();
+        assert_eq!(local_genesis, remote_genesis, "genesis blocks should match, we computed {local_genesis:#?} but geth computed {remote_genesis:#?}");
+
+        // set the chainspec genesis hash
+        chainspec.genesis_hash = local_genesis.hash();
+
+        // === create many blocks ===
+
+        let nonces = 0..1000u64;
+        let txs = nonces
+            .map(|nonce| {
+                // create a tx that just sends to the zero addr
+                TypedTransaction::Eip1559(Eip1559TransactionRequest::new()
+                    .to(H160::zero())
+                    .value(1u64)
+                    .nonce(nonce))
+            });
+        tracing::info!("generated transactions for blocks");
+
+        // finally send the txs to geth
+        provider.send_requests(txs).await.unwrap();
+
+        let block = provider.get_block_number().await.unwrap();
+        assert!(block > U64::zero());
+
+        // === initialize reth networking stack ===
+
+        let secret_key = SecretKey::new(&mut rand::thread_rng());
+        let (reth_p2p, reth_disc) = unused_tcp_udp();
+        tracing::info!(
+            %reth_p2p,
+            %reth_disc,
+            "setting up reth networking stack in keepalive test"
+        );
+
+        let config = NetworkConfig::<Arc<NoopProvider>>::builder(secret_key)
+            .listener_addr(reth_p2p)
+            .discovery_addr(reth_disc)
+            .chain_spec(chainspec.clone())
+            .build(Arc::new(NoopProvider::default()));
+
+        let network = NetworkManager::new(config).await.unwrap();
+        let handle = network.handle().clone();
+
+        tokio::task::spawn(network);
+
+        // create networkeventstream to get the next session established event easily
+        let mut events = NetworkEventStream::new(handle.event_listener());
+        let geth_socket = SocketAddr::new([127, 0, 0, 1].into(), geth_p2p_port);
+
+        // get the peer id we should be expecting
+        let geth_peer_id: PeerId = provider.peer_id().await.unwrap();
+
+        // add geth as a peer then wait for `PeerAdded` and `SessionEstablished` events.
+        handle.add_peer(geth_peer_id, geth_socket);
+
+        // wait for the session to be established
+        let _peer_id = events.peer_added_and_established().await.unwrap();
```

Comment:
shouldn't we add a `assert_eq!(geth_peer_id, peer_id)` here? what is the purpose of the keepalive stuff below?